### ISSUE TITLE
Issue more user-friendly errors on variable and parameter errors

### ISSUE THREAD

**Issue by vkarak** [eth-cscs/reframe#1899]
My example is particular for the required variables, but the title is more generic, since I believe we need check the error messages also in these cases.

If a variable is required or a parameter is abstract and the framework tries to instantiate the test, the user should not get a full stack trace unconditionally. As a user, when I see a stack trace, I am expecting a programming failure and not an error from the framework. If these errors were some type of `ReframeError`s, the user can get the stack trace by increasing the verbosity of the output.

**Comment by jjotero** [comment]
This is the current order on which these checks are made:

1. Check that parameters are defined: This is done in the `simple_test` decorator because the decorator itself needs the full length of the parameter space. At the moment, this throws an error and perhaps it would make sense to just skip the test.
2. Check the test defines `valid_systems` and `valid_prog_environs`: This is done by the loader, but it could also be done by the `simple_test` decorator. These two variables and the parameters are needed for the test's instantiation.
3. Check that any other variable is set: This is done when a variable is accessed during the test's execution. A required variable can be set at a later stage by a hook, before its value is needed.

The first two checks are the framework's responsibility and they should never be an `AttributeError`. I think that just skipping the tests in these scenarios and raising a warning would make sense. That being said, it might seem that points 1 & 2 could be moved into the `simple_test` decorator, but these two checks are conceptually different. A parameter determines how many instances are to be created from a class, and the variables `valid_systems` and `valid_prog_environs` determine whether a given instance is well-defined to be executed by reframe. The following valid scenario shows why the check done by point 2 should remain in the loader. 
```python
@rfm.simple_test
class MyTest(rfm.RegressionTest):
    executable = 'echo bananas'
    sanity_patterns = sn.assert_true(1)

MyTest.valid_systems._default_value = ['daint:login']
MyTest.valid_prog_environs._default_value = ['PrgEnv-gnu']
```
If the check on `valid_prog_environs` and `valid_systems` was done in the decorator, Reframe would skip this test despite that the test is good to execute.

Now onto point 3. If this check gets triggered, it's because of a programming error on the user side. Say the user does something like the following:
```python
@rfm.simple_test
class A(rfm.RunOnlyRegressionTest):
    executable = f'echo {v}'
    sanity_patterns = sn.assert_true(1)
```
This this is just Python and the user gets a `NameError` here. Now, say that the user has the following slight variation of the above example:
```python
@rfm.simple_test
class B(rfm.RunOnlyRegressionTest):
    v = variable(str)
    executable = f'echo {v}'
    sanity_patterns = sn.assert_true(1)
```
Here the user can't get a `NameError` because `v` has been declared. Instead, we throw a `ValueError` because the variable `v` has no value assigned to it. Now moving the class body from these two examples into hooks we get:
```python
@rfm.simple_test
class C(rfm.RunOnlyRegressionTest):
    @rfm.run_before('run')
    def set_executable_and_sanity(self):
        self.executable = f'echo {self.v}'
        self.sanity_patterns = sn.assert_true(1)

@rfm.simple_test
class D(rfm.RunOnlyRegressionTest):
    v = variable(str)
    @rfm.run_before('run')
    def set_executable_and_sanity(self):
        self.executable = f'echo {self.v}'
        self.sanity_patterns = sn.assert_true(1)
```
Test `C` is just Python again and the user will get an `AttributeError` because they have done a programming error. The same applies to `D`, where it's the user's fault not having defined `self.v` before it was required. Whether the variable `v` was declared or not is irrelevant, because this is always an error when accessing an attribute (i.e. an `AttributeError`). 

In short, if a user gets a regular Python error for `A` and `C`, they should also get a regular Python error for `B` and `D`, and never a `ReframeError`.

**Comment by ekouts** [comment]
I couldn't really find another case of getting a stack trace, when we shouldn't. I would suggest to just close this issue. @vkarak what do you think?

**Comment by jjotero** [comment]
@ekouts I think that with or without throwing a stack trace, I think that we should replace all the errors raised by the variables and parameters as custom errors (such as `ReframeSyntaxError` and so on) instead of the Python builtin errors. From the discussion during yesterday's standup, I think it would be great if we include a Rationale section in the docs explaining why sometimes we raise python builtin errors and why some other times we raise our custom exceptions.

**Comment by vkarak** [comment]
One thing we could improve with this issue, regardless of the variables/parameters is the standard error message when a test has problems:

```
./bin/reframe: skipping test 'HelloTest' due to errors: use `-v' for more information
    FILE: /Users/karakasv/Repositories/misc/rfm-support/issue-1899/hello1.py:19
```

The error here is `name 'x' is not defined`, but I would like to see it printed along with the skip message. Also, we could show the actual line after the `FILE` line.

**Comment by vkarak** [comment]
Actually, we don't need to change the type of error raised if we simply print the exception error message.

**Comment by vkarak** [comment]
Another problem is that error message above doesn't contain any file information if something happens at the class body. For example, the following:

```python
@rfm.simple_test
class HelloTest(rfm.RegressionTest):
    x = variable(int)
    x = 'foo'
```

issues this error:

```
./bin/reframe: skipping test 'HelloTest' due to errors: use `-v' for more information
    FILE: n/a:n/a
```

Due to the way variables and parameters are injected, we cannot currently show any file or lineno information, but this can be left for future work, but at least we could show the error message, which is `failed to set field 'x': 'foo' is not of type 'int'`.
